### PR TITLE
Rover land detection

### PR DESCRIPTION
**Pull request by ealdaz-seesai**
**Describe problem solved by this pull request**
In the current implementation when Land is requested the rover never lands but circles continuosly around the position where land was requested.
This is due to the fact that the RoveLandDetector currently returns the arm status, and once the Rover is armed it remains armed and hence land is never achieved.

**Describe your solution**
In the proposed implementation if landing is requested the Land detector detects that `vehicle_status.nav_state == vehicle_status_s::NAVIGATION_STATE_AUTO_LAND` 
and then simply returns  true.

**Test data / coverage**
This was tested with both px4_sitl gazebo_rover and a pixhawk4 rover running the Aion RObotics R1 UGV (Tank type/Differential drive)

**Current Behaviour**
Rover circles continously around landing position and landing is never detected
image

**Proposed Behaviour.** 
Rover stops and landing is detected.
image

### CODE REVIEW

File: src/modules/land_detector/RoverLandDetector.h [hunk 1]
```diff
@@ -56,7 +57,11 @@ class RoverLandDetector : public LandDetector
 	bool _get_ground_contact_state() override;
 	bool _get_landed_state() override;
 
+
 private:
+	// Program crashes when Subscriptor declared here
+	//uORB::Subscription _vehicle_status_sub{ORB_ID(vehicle_status)};
```
Comment (dagar) [inline src/modules/land_detector/RoverLandDetector.h]: Can you revert this change and we'll work through whatever issue you hit?

File: src/modules/land_detector/RoverLandDetector.cpp [hunk 2]
```diff
@@ -44,14 +44,31 @@
...
 
 bool RoverLandDetector::_get_landed_state()
 {
-	return !_actuator_armed.armed;
+
+// I need to do the uORB subscription here, can't do it in the .hpp file (SIGSEGV if I do... but no idea why)
+uORB::Subscription vehicle_status_sub{ORB_ID(vehicle_status)};
+vehicle_status_s vehicle_status{};		/**< vehicle status */
+
+vehicle_status_sub.update(&vehicle_status);
```
Comment (dagar) [inline src/modules/land_detector/RoverLandDetector.cpp]:
```suggestion
	_vehicle_status_sub.update(&_vehicle_status);
```

File: src/modules/land_detector/RoverLandDetector.cpp [hunk 3]
```diff
@@ -44,14 +44,31 @@
...
 	return true;
 }
 
 bool RoverLandDetector::_get_landed_state()
 {
-	return !_actuator_armed.armed;
+
+// I need to do the uORB subscription here, can't do it in the .hpp file (SIGSEGV if I do... but no idea why)
+uORB::Subscription vehicle_status_sub{ORB_ID(vehicle_status)};
+vehicle_status_s vehicle_status{};		/**< vehicle status */
```
Comment (dagar) [inline src/modules/land_detector/RoverLandDetector.cpp]:
```suggestion
```

File: src/modules/land_detector/RoverLandDetector.cpp [hunk 4]
```diff
@@ -44,14 +44,31 @@
...
 {
 	return true;
 }
 
 bool RoverLandDetector::_get_landed_state()
 {
-	return !_actuator_armed.armed;
+
+// I need to do the uORB subscription here, can't do it in the .hpp file (SIGSEGV if I do... but no idea why)
+uORB::Subscription vehicle_status_sub{ORB_ID(vehicle_status)};
```
Comment (dagar) [inline src/modules/land_detector/RoverLandDetector.cpp]:
```suggestion
```

File: src/modules/land_detector/RoverLandDetector.cpp [hunk 5]
```diff
@@ -44,14 +44,31 @@
...
+
+vehicle_status_sub.update(&vehicle_status);
+
+if(vehicle_status.nav_state == vehicle_status_s::NAVIGATION_STATE_AUTO_LAND)
+{
+	return true; // If Landing has been requested then say we have landed
+}
+else
+{
+	return !_actuator_armed.armed;  // If we are armed we are not landed
```
Comment (dagar) [inline src/modules/land_detector/RoverLandDetector.cpp]: The system being disarmed should probably be checked first to override.

File: src/modules/land_detector/RoverLandDetector.h (hunk 1 above)
Comment (ealdaz-seesai) [reply src/modules/land_detector/RoverLandDetector.h]: Hello dagar, apologies for delay, was away for Xmas.
Have just reverted the changes on my branch.
Please let me know if there is anything else I can help with

File: src/modules/land_detector/RoverLandDetector.cpp (hunk 5 above)
Comment (ealdaz-seesai) [reply src/modules/land_detector/RoverLandDetector.cpp]: Hello dagar, I don't really follow what you mean. This part of the code is unchanged.

Comment (dagar) [reply src/modules/land_detector/RoverLandDetector.cpp]: I mean you can check `_actuator_armed.armed` before updating vehicle_status and checking the nav_state.

**Comment by dagar** [comment]
CI is failing due to code style. 

http://ci.px4.io:8080/blue/organizations/jenkins/PX4%2FFirmware/detail/PR-13769/2/pipeline

http://ci.px4.io:8080/blue/organizations/jenkins/PX4%2FFirmware/detail/PR-13769/2/pipeline

File: src/modules/land_detector/RoverLandDetector.cpp (hunk 5 above)
Comment (ealdaz-seesai) [reply src/modules/land_detector/RoverLandDetector.cpp]: Ah, I think my comment in the code might be confusing things. The system works perfectly as it is currently.

The original part of the code (which is now part of the else) is what didn't work properly 
```
 else {
		return !_actuator_armed.armed;  // If we are armed we are not landed
		// This doesn't work very well because once it's armed it will stay armed
		// so when requesting land it never detects it has landed.
```
I'll remove those comments, as they are confusing

Comment (ealdaz-seesai) [reply src/modules/land_detector/RoverLandDetector.cpp]: Done

**Comment by julianoes** [comment]
@ealdaz-seesai I'm curious. How does your rover "land" :smile: ? Or what is the use-case to use land with a rover?

**Comment by julianoes** [comment]
@ItsTimmy would be great if you could review this one, thanks!

**Comment by ealdaz-seesai** [comment]
Ha,ha,ha ... yes, surprising concept for sure. In fact we use land as a proxy to end the mission in the same way as for a drone. We use both drones and rovers so it made sense to us to use the same commands for both.
Open to better ideas or suggestions

Thanks for following up on this pull request  as I was about to reach out and enquire if there was anything else I needed to do.

Eduardo

**Comment by ealdaz-seesai** [comment]
Is there anything else required to merge this?

**Comment by julianoes** [comment]
Thanks @ealdaz-seesai!

**Comment by ealdaz-seesai** [comment]
A pleasure, thank you!
